server/manageuser.php

```diff
@@ -0,0 +1,178 @@
+<?php
+
+require_once 'config.php';
+require_once "user.class.php";
+
+$USER = new User();
+
+if (!$USER->hasPermission('edit_user_permission')) {
+	header("Location: /unauthorized");
+	die();
+}
+
+if(isset($_REQUEST['callback'])){
+	switch ($_REQUEST['callback']) {
+		case 'getUserTable':
+			getUserTable($_REQUEST);
+			break;
+		case 'updateUserPermissions':
+			updateUserPermissions($_REQUEST);
+			break;
+		case 'addUser':
+			addUser($_REQUEST);
+			break;
+		default:
+			die('Unregistered command. This instance has been reported.');
+			break;
+	}
+}
+
+function getUserTable($data){
+	GLOBAL $DB_CONN;
+
+	$response = array();
+	$response['success'] = true;
+
+	$stmt = $DB_CONN->prepare("SELECT userId, firstName, lastName, email, preparesTaxes, archived 
+		FROM vita.user
+		ORDER BY firstName, lastName");
+
+	$stmt->execute(array());
+
+	$thead = "<thead><tr><th>Name</th><th>Email</th><th>Permissions</th><th>Count</th></tr>";
+	$tbody = "<tbody>";
+	while($row = $stmt->fetch(PDO::FETCH_ASSOC)){
+		$optionList = getUserPermissionOptionList($row['userId']);
+
+		$tbody.= "<tr data-user-id='".$row['userId']."'>";
+		$tbody.= "<td>".$row['firstName']." ".$row['lastName']."</td>";
+		$tbody.= "<td>".$row['email']."</td>";
+		$tbody.= "<td><select class='userPermissionList selectpicker' multiple=true>".implode('', $optionList['options'])."</select></td>";
+		$tbody.= "<td>".$optionList['hasPermissionCount']."</td>";
+		$tbody.= "</tr>";
+	}
+	$tbody.= "</tbody>";
+	$thead.= "</tr></thead>";
+
+	$response['table'] = $thead.$tbody;
+
+	print json_encode($response);
+}
+
+// Returns string of html <option> elments with all permissions
+function getUserPermissionOptionList($userId){
+	GLOBAL $DB_CONN;
+
+	$stmt = $DB_CONN->prepare("SELECT 
+		permissionId,
+		name,
+        description,
+        lookupName,
+        (SELECT userPermissionId FROM UserPermission WHERE userId = ? AND permissionId = p.permissionId) as userPermissionId
+	FROM Permission p");
+
+	$stmt->execute(array($userId));
+
+	$options = array();
+
+	$count = 0;
+	while($row = $stmt->fetch(PDO::FETCH_ASSOC)){
+		if($row['userPermissionId']){
+			$selected = 'selected=true';
+			$data = "data-userPermissionId='".$row['userPermissionId']."'";
+			$count++;
+		}else{
+			$selected = '';
+			$data = '';
```

Comment:
in the case where `$row['userPermissionId']` is not defined, do we really want to add something at all to this options array? Why not `continue` here? Or better yet, wrap it all in the `if($row['userPermissionId'])` block? I'm not understanding this else case.

Reply:
Yes, the options where that is not defined are the possible permissions they don't have yet. Un-selected options so to speak. If we don't build them in, then the user can't add new permissions to someone because they(the ungranted permissions) won't be in the dropdown.